server/src/main/java/org/elasticsearch/cluster/metadata/MetadataUpdateSettingsService.java

```diff
@@ -192,9 +198,56 @@ ClusterState execute(ClusterState currentState) {
-                );
+                if (request.reopenShards()) {
+                    // We have non-dynamic settings and open indices. We will unassign all of the shards in these indices so that the new
+                    // changed settings are applied when the shards are re-assigned.
+                    routingTableBuilder = RoutingTable.builder(
+                        allocationService.getShardRoutingRoleStrategy(),
+                        currentState.routingTable()
+                    );
+                    for (Index index : openIndices) {
+                        // We only want to take on the expense of reopening all shards for an index if the setting is really changing
+                        Settings existingSettings = currentState.getMetadata().index(index).getSettings();
+                        boolean needToReopenIndex = false;
+                        for (String setting : skippedSettings) {
+                            String newValue = request.settings().get(setting);
+                            if (Objects.equals(newValue, existingSettings.get(setting)) == false) {
+                                needToReopenIndex = true;
+                                break;
+                            }
+                        }
+                        if (needToReopenIndex) {
+                            List<ShardRouting> shardRoutingList = currentState.routingTable().allShards(index.getName());
+                            IndexRoutingTable.Builder indexRoutingTableBuilder = IndexRoutingTable.builder(index);
+                            for (ShardRouting shardRouting : shardRoutingList) {
+                                if (ShardRoutingState.UNASSIGNED.equals(shardRouting.state()) == false) {
+                                    indexRoutingTableBuilder.addShard(
+                                        shardRouting.moveToUnassigned(
+                                            new UnassignedInfo(
+                                                UnassignedInfo.Reason.INDEX_REOPENED,
+                                                "Unassigning shards to update static settings"
+                                            )
+                                        )
+                                    );
+                                } else {
+                                    indexRoutingTableBuilder.addShard(shardRouting);
+                                }
+                            }
+                            routingTableBuilder.add(indexRoutingTableBuilder.build());
+                            openIndices.remove(index);
+                            closedIndices.add(index);
+                        }
+                    }
+                } else {
+                    throw new IllegalArgumentException(
+                        String.format(
+                            Locale.ROOT,
+                            "Can't update non dynamic settings [%s] for open indices %s",
```

Comment:
It'd be good to make this message more actionable: suggest passing the `?reopen` query parameter (and also perhaps mention closing and reopening the index by hand)

Reply:
Done